src/vm/eventpipe.cpp

```diff
@@ -654,20 +672,31 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
                     payload.GetSize(),
                     pActivityId,
                     pRelatedActivityId);
-                instance.EnsureStack(*pSession);
+                instance.EnsureStack(*pRundownSession);
 
                 // EventPipeFile::WriteEvent needs to allocate a metadata event
                 // and can therefore throw. In this context we will silently
                 // fail rather than disrupt the caller
                 EX_TRY
                 {
-                    pSession->WriteEvent(instance);
+                    pRundownSession->WriteEvent(instance);
                 }
                 EX_CATCH {}
                 EX_END_CATCH(SwallowAllExceptions);
             }
         }
     }
+    else
+    {
+        for (Volatile<EventPipeSession *> &session : s_pSessions)
```

Comment:
We didn't quite get the right scope of protection when we discussed this the other day. The existing SuspendWriteEvent() ensure that threads won't create new buffer lists, buffers, or write events into those buffers. It does not do anything to prevent them enumerating this for loop or invoking pSession-WriteEvent() below. AVs remain possible. I have thoughts on how we can resolve this but I need to run to dinner. I'll write more later.

Reply:
Just to follow up on this one. I think my current best idea is to hoist the initialization of the EventPipeThread object outside this loop so that you can have a thread-local flag that is set the entire time the session pointer is in use. Something like this here:

```
EventPipeEventThread* pEventPipeThread = EventPipeThread::GetOrCreate();
for(int i = 0; i < MaxSessions; i++)
{
    // This read is OK because we aren't derefencing the pointer and if we observe a value that
    // isn't up-to-date (whether null or non-null) that is just the natural race timing of trying to
    // write to a session while it is being concurrently enabled/disabled.
    if(s_pSessions[i].LoadWithoutBarrier() == NULL)
        continue;

    // Now that we know this session is probably live we pay the perf cost of the memory barriers
    // Setting this flag lets a thread trying to do a concurrent disable that it is not safe to delete
    // session ID i. The if check above also ensures that once the session is unpublished this thread
    // will eventually stop ever storing ID i into the WriteInProgress flag. This is important to
    // guarantee termination of the YIELD_WHILE loop in SuspendWriteEvents.
    pEventPipeThread->VolatileSessionWriteInProgress.Store(i); 
    EventPipeSession* pSession = s_pSessions[i].Load();
    // The NULL check above may make this check below appear redundant but it is not. Disable is
    // allowed to set s_pSessions[i] = NULL at any time and that may have occured in between
    // the check and the load
    if(pSession != NULL)
    {
        // Write the event to the session
    }
    // Do not reference pSession past this point, we are signaling Disable() that it is safe to
    // delete it
    pEventPipeThread->VolatileSessionWriteInProgress.StoreWithoutBarrier(-1); 
}
```

In SuspendWriteEvents() 

```
// I don't believe this memory barrier is needed for correctness, but given effectively no perf 
// constraint in Disable() there is no reason to be stingy
// Setting this to NULL stops the inflow of new threads trying to write to this session but we
// still need to assume that some threads already have writes in progress
s_pSessions[this_session_id].Store(NULL); 

// the rest of the SuspendWriteEvents code...

// replace the YIELD_WHILE condition with:
YIELD_WHILE(pEventPipeThread->VolatileSessionWriteInProgress.Load() == this_session_id)
// It still guarantees that the thread has returned its buffer, but it also now guarantees that
// that the thread has returned from Session::WriteEvent() and has relinquished the session
// pointer
// This yield is guaranteed to eventually finish because threads will eventually exit WriteEvent()  
// setting the flag back to -1. If the thread could quickly re-enter WriteEvent and set the flag
// back to this_session_id we could theoretically get unlucky and never observe the gap, but 
// setting s_pSessions[this_session_id] = NULL above guaranteed that can't happen indefinately.
// Sooner or later the thread is going to see the NULL value and once it does it won't store
// this_session_id into the flag again.
// As above I'm pretty sure this read barrier isn't necessary for correctness, but this isn't the hot
// path so I am not being stingy
```

In Disable()
```
// after this call returns we are guaranteed that no thread will try to dereference the session 
// pointer in write event (nor any of the memory such as buffer manager, buffer list, buffer)
SuspendWriteEvents()

// do rundown which uses the session on this thread only

delete pSession;

// the session id and slot in the s_pSessions is now safe for reuse
```